validator-core - add hability to load data to PostgreSQL in a custom schema

Database connection is set from environment variables.

- VALIDATOR_DB_URL 
- VALIDATOR_DB_USER
- VALIDATOR_DB_PASSWORD
- VALIDATOR_DB_SCHEMA

Good job : 

Version changed to "4.0.0-SNAPSHOT" (3.4 branch for GpU 3.3.x), break change incoming in metadata (identifier -> identifiers) for GpU 4.x (please wait before release)

Maybe :

* Remove VALIDATOR_ prefix to avoid duplicating `DB_URL`, `DB_USER`, `DB_PASSWORD` already defined for other projects (dsr)

So far : 

```bash
export VALIDATOR_DB_URL=jdbc:postgresql:validator-test
export VALIDATOR_DB_USER=postgis
export VALIDATOR_DB_PASSWORD=postgis
# provides a nice message : PSQLException: FATAL: database "validator-test" does not exist
mvn clean package
# and this works globally works (some test to be fixed)
createdb validator-test
psql -d validator-test -c "CREATE EXTENSION postgis"
mvn clean package
```

```
[ERROR] Failures: 
[ERROR]   DgprApplicationTest.testDocument:94 expected:<0> but was:<46>
[ERROR]   DgprApplicationTest.testDocumentNotOk:273 expected:<2> but was:<0>
[ERROR]   DgprApplicationTest.testDocumentNotOkTolerance10noSafe:184 expected:<2> but was:<0>
[ERROR]   DgprApplicationTest.testDocumentNotOkTolerance10safe:217 expected:<2> but was:<0>
[ERROR]   DgprApplicationTest.testDocumentOkTolerance5:123 expected:<0> but was:<40>
[ERROR]   DgprApplicationTest.testDocumentOkWithNoTolerance:244 expected:<0> but was:<12>
```

--- a/validator-core/src/main/java/fr/ign/validator/database/Database.java
+++ b/validator-core/src/main/java/fr/ign/validator/database/Database.java
@@ -12,6 +12,7 @@ import java.sql.Statement;
 import java.util.ArrayList;
 import java.util.List;
 
+import org.apache.commons.lang.StringUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.logging.log4j.Marker;
@@ -42,18 +43,16 @@ import fr.ign.validator.tools.TableReader;
 public class Database {
 
     public static final Logger log = LogManager.getRootLogger();
-    public static final Marker MARKER = MarkerManager.getMarker("DocumentDatabase");
+    public static final Marker MARKER = MarkerManager.getMarker("Database");
 
     private static final int batchSize = 100;
 
-    public static final String ENV_DATABASE_URL = "DB_URL";
-    public static final String ENV_DATABASE_USER = "DB_USER";
-    public static final String ENV_DATABASE_PASSWORD = "DB_PASSWORD";
-    public static final String ENV_DATABASE_SCHEMA = "DB_SCHEMA";
+    private static final String ENV_DATABASE_URL = "DB_URL";
+    private static final String ENV_DATABASE_USER = "DB_USER";
+    private static final String ENV_DATABASE_PASSWORD = "DB_PASSWORD";
+    private static final String ENV_DATABASE_SCHEMA = "DB_SCHEMA";
 
     public static final String POSTGRESQL_DRIVER = "PostgreSQL Native Driver";
-
-    public static final String GEOMETRY_COLUMN = "WKT";
 
     public static final String DEFAULT_SRID = "4326";
 
@@ -67,7 +66,7 @@ public class Database {
     private Connection connection;
 
     /**
-     * Create or open an sqlite database
+     * Create or open an SQLITE database
      * 
      * @param sqlitePath
      */
@@ -85,24 +84,31 @@ public class Database {
         }
     }
 
-    public Database(String url, String user, String password, String schema) {
-        log.info(MARKER, "Create POSTGRESQL database {}...", url);
+    /**
+     * Create or open a PostgreSQL database with a given schema
+     * 
+     * @param url
+     * @param user
+     * @param password
+     * @param schema
+     */
+    private Database(String url, String user, String password, String schema) {
+        log.info(MARKER, "Create PostgreSQL database {} using schema={}...", url, schema);
         try {
             Class.forName("org.postgresql.Driver");
             connection = DriverManager.getConnection(url, user, password);
             connection.setAutoCommit(false);
-            // save schema
-            this.setSchema(schema);
+            this.schema = schema;
+            updateCurrentSchema(false);
         } catch (ClassNotFoundException e) {
             throw new RuntimeException(e);
         } catch (SQLException e) {
             throw new RuntimeException(e);
         }
-
     }
 
     /**
-     * Get database connection
+     * Get SQL connection
      * 
      * @return
      */
@@ -111,71 +117,126 @@ public class Database {
     }
 
     /**
-     * Create database with a schema corresponding to the given document
+     * True if a schema is defined
+     */
+    private boolean hasSchema() {
+        return !StringUtils.isEmpty(this.schema);
+    }
+
+    /**
+     * If defined, ensure that the specified schema exists and set in search_path.
      * 
+     * @param reset true if the schema has to be re-created
+     * @throws SQLException
+     */
+    private void updateCurrentSchema(boolean reset) throws SQLException {
+        if (!hasSchema()) {
+            return;
+        }
+        // recreate schema is reset is set to true
+        if (reset) {
+            update("DROP SCHEMA IF EXISTS " + schema + " CASCADE;");
+        }
+        // ensure that schema exists
+        update("CREATE SCHEMA IF NOT EXISTS " + schema + ";");
+        // set current search path to use current schema
+        update("SET search_path = " + schema + ", public;");
+        connection.commit();
+    }
+
+    public Projection getProjection() {
+        return projection;
+    }
+
+    public void setProjection(Projection projection) {
+        this.projection = projection;
+    }
+
+    /**
+     * Create a validation Database according to environment variables :
+     * 
+     * <ul>
+     * <li>DB_URL (ex : "jdbc:postgresql:validator-test")</li>
+     * <li>DB_USER (ex : "postgis")</li>
+     * <li>DB_PASSWORD (ex : "postgis")</li>
+     * <li>DB_SCHEMA (ex : "validation_{document.id}")</li>
+     * </ul>
+     * 
+     * Note that :
+     * 
+     * <ul>
+     * <li>An SQLITE database is created in validation directory if DB_URL is not
+     * specified</li>
+     * <li>The default schema for PostgreSQL is "validation" and "public" is
+     * forbidden</li>
+     * <li>DB_SCHEMA is set a the default schema (search path)
+     * <li>
+     * </ul>
+     *
      * @param document
+     * @param reset    true if the schema has to be re-created
+     * 
      * @return
      * @throws SQLException
      */
-    public static Database createDatabase(Document document) throws SQLException {
+    public static Database createDatabase(Context context, boolean reset) throws SQLException {
         String url = System.getenv(Database.ENV_DATABASE_URL);
-        if (url != null && !url.isEmpty()) {
-            return createPostgresDatabase(document);
+        if (StringUtils.isEmpty(url)) {
+            return Database.createSqlLiteDatabase(context, reset);
+        } else {
+            return createPostgresDatabase(context, reset);
         }
-        // else
-        return Database.createSqlLiteDatabase(document);
     }
 
-    private static Database createSqlLiteDatabase(Document document) throws SQLException {
-        File parentDirectory = document.getDocumentPath().getParentFile();
-        File databasePath = new File(parentDirectory, "document_database.db");
-        if (databasePath.exists()) {
-            log.info(MARKER, "remove existing database {}...", databasePath);
+    /**
+     * Create an SQLITE database as document_database.db in validation directory.
+     * 
+     * @param context
+     * @param reset
+     * @return
+     * @throws SQLException
+     */
+    private static Database createSqlLiteDatabase(Context context, boolean reset) throws SQLException {
+        File databasePath = new File(context.getValidationDirectory(), "document_database.db");
+        if (databasePath.exists() && reset) {
+            log.info(MARKER, "Remove existing database {}...", databasePath);
             databasePath.delete();
         }
-        Database database = new Database(databasePath);
-        if (database.hasSchema()) {
-            database.createSchema(database.getSchema());
-        }
-        database.createTables(document.getDocumentModel());
-        return database;
+        return new Database(databasePath);
     }
 
-    private static Database createPostgresDatabase(Document document) throws SQLException {
+    /**
+     * Create a PostgreSQL database according to environment variables.
+     * 
+     * @param context
+     * @param reset
+     * @return
+     * @throws SQLException
+     */
+    private static Database createPostgresDatabase(Context context, boolean reset) throws SQLException {
         String url = System.getenv(Database.ENV_DATABASE_URL);
         String user = System.getenv(Database.ENV_DATABASE_USER);
         String password = System.getenv(Database.ENV_DATABASE_PASSWORD);
         String schema = System.getenv(Database.ENV_DATABASE_SCHEMA);
-        Database database = new Database(url, user, password, schema);
-        if (database.hasSchema()) {
-            database.createSchema(database.getSchema());
+        if (StringUtils.isEmpty(schema)) {
+            schema = "validation";
+        } else if (schema.equals("public")) {
+            throw new RuntimeException("The use DB_SCHEMA=public is forbidden");
         }
-        database.createTables(document.getDocumentModel());
+        Database database = new Database(url, user, password, schema);
+        if (reset) {
+            database.updateCurrentSchema(true);
+        }
         return database;
     }
 
-    public void createSchema(String schemaName) throws SQLException {
-        String dropClause = "";
-        if (this.isPostgresqlDriver() && !schemaName.equals("public")) {
-            dropClause = "DROP SCHEMA IF EXISTS " + schemaName + " CASCADE;";
-        }
-        String sql = dropClause
-            + " CREATE SCHEMA IF NOT EXISTS " + schemaName + ";"
-            + " SET search_path = " + schemaName + ", public;";
-
-        // debug SQL
-        log.debug(MARKER, sql);
-        Statement sth = connection.createStatement();
-        sth.executeUpdate(sql);
-        connection.commit();
-    }
-
     /**
-     * Create tables according to documentModel
+     * Create tables according to the TableModels of the DocumentModel
      * 
      * @throws SQLException
      */
     public void createTables(DocumentModel documentModel) throws SQLException {
+        log.info(MARKER, "Create tables for the DocumentModel '{}' ...", documentModel.getName());
         List<FileModel> fileModels = documentModel.getFileModels();
         for (FileModel fileModel : fileModels) {
             if (fileModel instanceof TableModel) {
@@ -186,12 +247,13 @@ public class Database {
     }
 
     /**
-     * Create table according to tableModel
+     * Create a table according to a TableModel
      * 
      * @param tableModel
      * @throws SQLException
      */
     public void createTable(TableModel tableModel) throws SQLException {
+        log.info(MARKER, "Create table for the TableModel '{}' ...", tableModel.getName());
         if (this.isPostgresqlDriver()) {
             createPostgisTable(tableModel);
         } else {
@@ -200,7 +262,7 @@ public class Database {
     }
 
     /**
-     * Create table
+     * Create a table with a list of text columns.
      * 
      * @param string
      * @param columnNames
@@ -219,10 +281,7 @@ public class Database {
         }
         sql += ");";
 
-        // debug SQL
-        log.debug(MARKER, sql);
-        Statement sth = connection.createStatement();
-        sth.executeUpdate(sql);
+        update(sql);
         connection.commit();
     }
 
@@ -232,13 +291,17 @@ public class Database {
      * @param tableModel
      * @throws SQLException
      */
-    public void createPostgisTable(TableModel tableModel) throws SQLException {
+    private void createPostgisTable(TableModel tableModel) throws SQLException {
         List<AttributeType<?>> attributes = tableModel.getFeatureType().getAttributes();
         List<String> columns = new ArrayList<String>(attributes.size());
 
         for (AttributeType<?> attribute : attributes) {
             if (attribute.isGeometry()) {
                 columns.add(attribute.getName() + " TEXT");
+                /*
+                 * TODO clarify "the_geom" (supporting "the_geom" as a default geom instead of
+                 * "WKT" would be better than this "hack")
+                 */
                 columns.add("the_geom geometry(" + attribute.getTypeName() + "," + Database.DEFAULT_SRID + ")");
             } else {
                 columns.add(attribute.getName() + " TEXT");
@@ -249,15 +312,29 @@ public class Database {
             + String.join(",", columns)
             + ");";
 
-        // debug SQL
-        log.debug(MARKER, sql);
-        Statement sth = connection.createStatement();
-        sth.executeUpdate(sql);
+        update(sql);
         connection.commit();
-        /*
-         * -- Add a spatial index CREATE INDEX mydata_geom_idx ON mydata USING gist
-         * (geom);
-         */
+    }
+
+    /**
+     * Create indexes for all tables on unique fields
+     * 
+     * @param document
+     * @throws SQLException
+     */
+    public void createIndexes(DocumentModel documentModel) throws SQLException {
+        for (FileModel file : documentModel.getFileModels()) {
+            if (!(file instanceof TableModel)) {
+                continue;
+            }
+            List<AttributeType<?>> attributes = file.getFeatureType().getAttributes();
+            for (AttributeType<?> attributeType : attributes) {
+                if (!attributeType.getConstraints().isUnique()) {
+                    continue;
+                }
+                createIndex(file.getName(), attributeType.getName());
+            }
+        }
     }
 
     /**
@@ -276,10 +353,8 @@ public class Database {
         String indexName = "idx_" + tableName + "_" + columnName;
         String sql = "CREATE INDEX " + ifClause + indexName
             + " ON " + tableName + " (" + columnName + ")";
-        // debug SQL
-        log.debug(MARKER, sql);
-        Statement sth = connection.createStatement();
-        sth.executeUpdate(sql);
+
+        update(sql);
         connection.commit();
     }
 
@@ -342,11 +417,10 @@ public class Database {
                 // must split code
                 srid = this.getProjection().getCode().split(":")[1];
             }
+            // TODO clarify wkt/the_geom
             String updateSQL = "UPDATE " + featureType.getName() + " SET the_geom = "
                 + "ST_Multi(ST_Transform(ST_SetSRID(wkt, " + srid + "), 4326));";
-            log.debug(MARKER, updateSQL);
-            Statement sth = connection.createStatement();
-            sth.executeUpdate(updateSQL);
+            update(updateSQL);
             connection.commit();
         }
     }
@@ -435,7 +509,7 @@ public class Database {
     }
 
     /**
-     * Perform any SQL request to DocumentDatabase
+     * Perform any SQL request returning results
      * 
      * @param sql
      * @return
@@ -452,6 +526,19 @@ public class Database {
     }
 
     /**
+     * Perform any SQL request that doesn't returns results
+     * 
+     * @param sql
+     * @return
+     * @throws SQLException
+     */
+    private void update(String sql) throws SQLException {
+        log.debug(MARKER, sql);
+        Statement sth = connection.createStatement();
+        sth.executeUpdate(sql);
+    }
+
+    /**
      * Return an Array of String giving column names for a giving table
      * 
      * @param tablename
@@ -459,7 +546,7 @@ public class Database {
      * @throws SQLException
      * @throws IOException
      */
-    public String[] getTableSchema(String tablename) throws SQLException, IOException {
+    private String[] getTableSchema(String tablename) throws SQLException, IOException {
         RowIterator query = query("SELECT * FROM " + tablename + " LIMIT 1");
         String[] header = query.getHeader();
         query.close();
@@ -467,7 +554,7 @@ public class Database {
     }
 
     /**
-     * Return the number of features of a giving table
+     * Return the number of rows for a given table
      * 
      * @param tableName
      * @return
@@ -481,36 +568,12 @@ public class Database {
     }
 
     /**
-     * Return all the feature of a giving table
+     * Returns true if driver is postgresql
      * 
-     * @param tablename
+     * TODO move to private, add helper to retrieve values
+     * 
      * @return
-     * @throws SQLException
      */
-    public RowIterator selectAll(String tablename) throws SQLException {
-        return query("SELECT * FROM " + tablename);
-    }
-
-    public boolean hasSchema() {
-        return this.schema != null && !this.schema.isEmpty();
-    }
-
-    public String getSchema() {
-        return schema;
-    }
-
-    public void setSchema(String schema) {
-        this.schema = schema;
-    }
-
-    public Projection getProjection() {
-        return projection;
-    }
-
-    public void setProjection(Projection projection) {
-        this.projection = projection;
-    }
-
     public boolean isPostgresqlDriver() {
         try {
             return connection.getMetaData().getDriverName().equals(Database.POSTGRESQL_DRIVER);

--- a/validator-core/src/test/java/fr/ign/validator/database/DatabaseTest.java
+++ b/validator-core/src/test/java/fr/ign/validator/database/DatabaseTest.java
@@ -1,21 +1,25 @@
 package fr.ign.validator.database;
 
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
 import java.io.File;
+import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.List;
 
 import org.apache.commons.io.FileUtils;
 import org.junit.Assert;
-import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
 
 import fr.ign.validator.Context;
 import fr.ign.validator.data.Document;
+import fr.ign.validator.io.JsonModelReader;
 import fr.ign.validator.io.ModelReader;
-import fr.ign.validator.io.XmlModelReader;
 import fr.ign.validator.model.DocumentModel;
 import fr.ign.validator.report.InMemoryReportBuilder;
 import fr.ign.validator.tools.ResourceHelper;
@@ -25,30 +29,25 @@ public class DatabaseTest {
     @Rule
     public TemporaryFolder folder = new TemporaryFolder();
 
-    InMemoryReportBuilder reportBuilder = new InMemoryReportBuilder();
-
-    Context context;
-
-    @Before
-    public void setUp() {
-        context = new Context();
-    }
-
     /**
-     * @throws Exception
+     * Create an SQLITE database with a specific path and performs basic tests
      */
     @Test
-    public void testEmptyDatabaseFromFile() {
+    public void testCreateDatabaseFile() {
         try {
-            // test database driver
-            File path = new File(folder.getRoot(), "document_database.db");
-            Database database = new Database(path);
+            File databaseFile = new File(folder.getRoot(), "sample.db");
+            Database database = new Database(databaseFile);
+
+            // ensure that file is created
+            assertTrue(databaseFile.exists());
+
+            // ensure that simple select works without any table
             RowIterator it = database.query("SELECT 'test' as test");
-            Assert.assertTrue(it.hasNext());
+            assertTrue(it.hasNext());
             String[] row = it.next();
-            Assert.assertEquals(1, row.length);
-            Assert.assertEquals("test", row[0]);
-            Assert.assertFalse(it.hasNext());
+            assertEquals(1, row.length);
+            assertEquals("test", row[0]);
+            assertFalse(it.hasNext());
         } catch (Exception e) {
             e.printStackTrace();
             Assert.fail();
@@ -56,24 +55,24 @@ public class DatabaseTest {
     }
 
     /**
-     * @throws Exception
+     * Performs basic test with some queries
      */
     @Test
     public void testCreateInsertSelect() {
         try {
-            File path = new File(folder.getRoot(), "document_database.db");
-            Database database = new Database(path);
+            File databaseFile = new File(folder.getRoot(), "sample.db");
+            Database database = new Database(databaseFile);
             database.query("CREATE TABLE TEST(id TEXT, name TEXT);");
             database.query("INSERT INTO TEST(id, name) VALUES ('1', 'name01');");
 
             RowIterator iterator = database.query("SELECT * FROM TEST;");
-            Assert.assertTrue(iterator.hasNext());
+            assertTrue(iterator.hasNext());
             int indexId = iterator.getColumn("id");
             int indexName = iterator.getColumn("name");
 
             String[] feature = iterator.next();
-            Assert.assertEquals("1", feature[indexId]);
-            Assert.assertEquals("name01", feature[indexName]);
+            assertEquals("1", feature[indexId]);
+            assertEquals("name01", feature[indexName]);
 
             iterator.close();
         } catch (Exception e) {
@@ -82,79 +81,11 @@ public class DatabaseTest {
         }
     }
 
-    protected Document getDocument(boolean validate) throws Exception {
-        context.setProjection("EPSG:4326");
-        context.setReportBuilder(reportBuilder);
-
-        File documentModelPath = ResourceHelper.getResourceFile(getClass(), "/config-xml/cnig_PLU_2014/files.xml");
-        ModelReader modelLoader = new XmlModelReader();
-        DocumentModel documentModel = modelLoader.loadDocumentModel(documentModelPath);
-
-        File documentPath = ResourceHelper.getResourceFile(getClass(), "/documents/41003_PLU_20130903");
-        File copy = folder.newFolder(documentPath.getName());
-        FileUtils.copyDirectory(documentPath, copy);
-
-        Document document = new Document(documentModel, copy);
-
-        File validationDirectory = new File(copy.getParentFile(), "validation");
-        context.setValidationDirectory(validationDirectory);
-
-        if (validate) {
-            document.validate(context);
-        }
-
-        return document;
-    }
-
     /**
+     * Load DUMMY.csv file with A,B,WKT columns in table with A,B columns.
+     * 
      * @throws Exception
      */
-    @Test
-    public void testDatabaseFromNonValidatedDocument() {
-        try {
-            Document document = getDocument(false);
-            Database database = Database.createDatabase(document);
-            database.load(context, document);
-
-            // no data will be load (because no validation, no csv, no file mapping
-            int count = database.getCount("DOC_URBA");
-            Assert.assertEquals(0, count);
-
-            int countPrescription = database.getCount("PRESCRIPTION_SURF");
-            Assert.assertEquals(0, countPrescription);
-
-        } catch (Exception e) {
-            e.printStackTrace();
-            Assert.fail();
-        }
-    }
-
-    /**
-     * @throws Exception
-     */
-    @Test
-    public void testDatabaseFromValidatedDocument() {
-        try {
-            Document document = getDocument(true);
-
-            Database database = Database.createDatabase(document);
-            database.load(context, document);
-
-            // some feature were load
-            // 0 DOC_URBA (because no DOC_URBA table
-            int count = database.getCount("DOC_URBA");
-            Assert.assertEquals(0, count);
-            // 19 PRESCRIPTION_SURF_41003
-            // 19 or 38 ? TODO figure how many...
-            int countPrescription = database.getCount("PRESCRIPTION_SURF");
-            Assert.assertEquals(19, countPrescription);
-
-        } catch (Exception e) {
-            e.printStackTrace();
-            Assert.fail(e.getMessage());
-        }
-    }
-
     @Test
     public void testLoadSimpleFileWithColumnsAandB() throws Exception {
         Database database = new Database(new File(folder.getRoot(), "test.sqlite"));
@@ -171,7 +102,9 @@ public class DatabaseTest {
     }
 
     /**
-     * Test insert file
+     * Load DUMMY.csv file with A,B,WKT columns in table with C,D columns.
+     * 
+     * @throws Exception
      */
     @Test
     public void testLoadFileWithNoMatchingColumns() throws Exception {
@@ -186,6 +119,57 @@ public class DatabaseTest {
         database.loadFile("test", file, StandardCharsets.UTF_8);
 
         Assert.assertEquals(0, database.getCount("test"));
+    }
+
+    @Test
+    public void testLoadAdresseMultiple() throws Exception {
+        Context context = createTestContext();
+
+        Document document = getSampleDocument("adresse", "adresse-multiple");
+        Database database = Database.createDatabase(context, true);
+        database.createTables(document.getDocumentModel());
+        /*
+         * Note that this line is required to find mapping between DocumentFiles and
+         * FileModels
+         */
+        document.findFileModelForFiles(context);
+        database.load(context, document);
+        assertEquals(8, database.getCount("adresse"));
+    }
+
+    /**
+     * Get sample Document
+     * 
+     * @return
+     * @throws IOException
+     */
+    protected Document getSampleDocument(String documentModelName, String documentName) throws IOException {
+        File documentModelPath = ResourceHelper.getResourceFile(
+            getClass(), "/config-json/" + documentModelName + "/files.json"
+        );
+        ModelReader modelLoader = new JsonModelReader();
+        DocumentModel documentModel = modelLoader.loadDocumentModel(documentModelPath);
+
+        File documentPath = ResourceHelper.getResourceFile(getClass(), "/documents/" + documentName);
+        File copy = folder.newFolder(documentPath.getName());
+        FileUtils.copyDirectory(documentPath, copy);
+        return new Document(documentModel, copy);
+    }
+
+    /**
+     * Create a test context
+     * 
+     * @return
+     */
+    private Context createTestContext() {
+        InMemoryReportBuilder reportBuilder = new InMemoryReportBuilder();
+        Context context = new Context();
+        context.setProjection("EPSG:4326");
+        File validationDirectory = new File(folder.getRoot(), "validation");
+        validationDirectory.mkdirs();
+        context.setValidationDirectory(validationDirectory);
+        context.setReportBuilder(reportBuilder);
+        return context;
     }
 
 }

--- a/validator-plugin-dgpr/src/main/java/fr/ign/validator/dgpr/database/ValidatableDatabase.java
+++ b/validator-plugin-dgpr/src/main/java/fr/ign/validator/dgpr/database/ValidatableDatabase.java
@@ -1,6 +1,5 @@
 package fr.ign.validator.dgpr.database;
 
-import java.sql.SQLException;
 import java.util.ArrayList;
 import java.util.List;
 
@@ -12,15 +11,12 @@ import org.apache.logging.log4j.MarkerManager;
 import fr.ign.validator.Context;
 import fr.ign.validator.data.Document;
 import fr.ign.validator.database.Database;
-import fr.ign.validator.model.AttributeType;
-import fr.ign.validator.model.FileModel;
-import fr.ign.validator.model.file.TableModel;
 import fr.ign.validator.validation.Validatable;
 import fr.ign.validator.validation.Validator;
 
 /**
  * 
- * @author sai
+ * @author CBouche
  */
 public class ValidatableDatabase implements Validatable {
 
@@ -31,11 +27,6 @@ public class ValidatableDatabase implements Validatable {
      * Database
      */
     private Database database;
-
-    /**
-     * Document
-     */
-    private Document document;
 
     /**
      * Validators
@@ -50,40 +41,17 @@ public class ValidatableDatabase implements Validatable {
      * @throws Exception
      */
     public ValidatableDatabase(Context context, Document document) throws Exception {
-        this.document = document;
-        this.database = Database.createDatabase(document);
-        createIndexes(document);
+        log.info(MARKER, "Create ValidatableDatabase...");
+        this.database = Database.createDatabase(context, true);
+        this.database.createTables(document.getDocumentModel());
+        this.database.createIndexes(document.getDocumentModel());
         database.setProjection(context.getProjection());
         database.load(context, document);
     }
 
-    /**
-     * create all table indexes on identifier fields
-     * 
-     * @param document
-     * @throws SQLException
-     */
-    private void createIndexes(Document document) throws SQLException {
-        for (FileModel file : document.getDocumentModel().getFileModels()) {
-            if (!(file instanceof TableModel)) {
-                continue;
-            }
-            List<AttributeType<?>> attributes = file.getFeatureType().getAttributes();
-            for (AttributeType<?> attributeType : attributes) {
-                if (!attributeType.isIdentifier()) {
-                    continue;
-                }
-                database.createIndex(file.getName(), attributeType.getName());
-            }
-        }
-    }
-
     @Override
     public void validate(Context context) throws Exception {
-        log.info(
-            MARKER, "Validation de la database {}",
-            document.getDocumentModel().getName()
-        );
+        log.info(MARKER, "Validate using database validators...");
 
         /*
          * Validation at document level
@@ -91,7 +59,6 @@ public class ValidatableDatabase implements Validatable {
         for (Validator<Database> validator : getValidators()) {
             validator.validate(context, database);
         }
-
     }
 
     /**
@@ -101,15 +68,6 @@ public class ValidatableDatabase implements Validatable {
      */
     public List<Validator<Database>> getValidators() {
         return validators;
-    }
-
-    /**
-     * set list of database validators
-     * 
-     * @param validators
-     */
-    public void setValidators(List<Validator<Database>> validators) {
-        this.validators = validators;
     }
 
     /**

--- a/validator-plugin-dgpr/src/main/java/fr/ign/validator/dgpr/validation/database/IdentifierValidator.java
+++ b/validator-plugin-dgpr/src/main/java/fr/ign/validator/dgpr/validation/database/IdentifierValidator.java
@@ -15,7 +15,6 @@ import fr.ign.validator.database.RowIterator;
 import fr.ign.validator.dgpr.error.DgprErrorCodes;
 import fr.ign.validator.error.ErrorScope;
 import fr.ign.validator.model.AttributeType;
-import fr.ign.validator.model.FeatureType;
 import fr.ign.validator.model.FileModel;
 import fr.ign.validator.model.file.TableModel;
 import fr.ign.validator.validation.Validator;
@@ -62,22 +61,25 @@ public class IdentifierValidator implements Validator<Database> {
             if (!(fileModel instanceof TableModel)) {
                 continue;
             }
-
-            FeatureType featureType = fileModel.getFeatureType();
-
-            List<AttributeType<?>> attributesList = featureType.getAttributes();
+            
+            // TODO context.beginModel(fileModel)
 
             // Looking for attributes who are identifiers
-            for (AttributeType<?> attribute : attributesList) {
+            for (AttributeType<?> attribute : fileModel.getFeatureType().getAttributes()) {
                 if (!attribute.isIdentifier()) {
                     continue;
                 }
+                // TODO context.beginModel(attribute)
                 validateOneIdentifier(attribute.getName(), fileModel);
+                // TODO context.endModel(attribute)
             }
+            // TODO context.endModel(fileModel)
         }
     }
 
     private void validateOneIdentifier(String identifier, FileModel fileModel) throws SQLException, IOException {
+        // TODO HAVING count(*) > 1 ORDER BY count(*) DESC
+        // TODO LIMIT 10 ?
         RowIterator table = database.query(
             "SELECT " + identifier + " AS id, Count(" + identifier + ") AS count "
                 + " FROM " + fileModel.getName()
@@ -92,6 +94,7 @@ public class IdentifierValidator implements Validator<Database> {
             int compte = Integer.parseInt(row[indexCount]);
 
             // if not unique, send error
+            // remove condition by adding HAVING count(*) > 1
             if (compte > 1) {
                 context.report(
                     context.createError(DgprErrorCodes.DGPR_IDENTIFIER_UNICITY)

